회원가입으로 Firebase 사용자 등록 및 코드 리팩토링

song1900 opened the pull request:
- CloseButton, BasicLabel 클래스를 추가했습니다.
- 회원가입 폼에 사용자 경험 향상을 위한 기능을 추가했습니다. 
- 회원가입 화면에 다국어 지원을 적용했습니다.
- Firebase 사용자 등록 기능을 추가했습니다.
- 중복 코드, 주석 작성 등 코드 가독성을 향상시키는 리팩토링을 진행했습니다.

song1900 committed:
Feat: CloseButton 클래스 추가 및 적용

앱 전체에서 닫기 버튼의 일관성을 유지하기 위해 CloseButton 커스텀 클래스를 추가했습니다.
버튼을 커스텀 이미지로 설정하고 테마(라이트/다크 모드)에 따라 컬러를 조정합니다.
회원가입 화면에 적용했습니다.

song1900 committed:
Refactor: 클래스에 final 키워드 추가

클래스 확장을 방지하기 위해 클래스에 final 키워드를 추가했습니다.

song1900 committed:
Refactor: 회원가입 화면 TextField가 빈값이면 BorderColor를 default값으로 변경

song1900 committed:
Feat: 텍스트 필드 위에 라벨을 추가하여 사용자 경험 향상

각 텍스트 필드가 포커스될 때 라벨이 보이도록 구현했습니다.

song1900 committed:
Merge branch 'main' of https://github.com/f-lab-edu/gil-navi-ios into feature/signup-registration

# Conflicts:
#	GIL/GIL/Component/BasicButton.swift
#	GIL/GIL/Screen/SignUp/SignUpView.swift

song1900 committed:
Chore: BasicTextField verfiyPassword type 추가

song1900 committed:
Chore: 회원가입 화면 다국어 적용 및 수정

song1900 committed:
Feat: BasicLabel 추가

song1900 committed:
Feat: 회원가입 비밀번호 확인

비밀번호의 조건들을 Label로 추가하고, 조건에 충족되면 Color값을 변경하도록 했습니다.

song1900 committed:
Feat: FirebaseAuth 확장, 에러 케이스 추가

- Firebase 사용자 생성 기능을 비동기적으로 처리하는 확장 메서드를 추가했습니다.
- FirebaseAuthError 열거형을 추가하여 인증 과정에서 발생할 수 있는 예상치 못한 에러를 처리하도록 했습니다.

song1900 committed:
Refactor: SignUpViewController, SignUpViewModel 리팩토링

- ViewController의 bindPublishers() 메서드에 여러 기능이 혼합되어 있어 가독성이 떨어져 각 기능별로 메소드를 나누는 작업을 진행했습니다.
- Combine 사용시 스레드처리가 미흡한 부분이 있어 수정했습니다.
- UI를 업데이트 하는 중복된 코드를 수정했습니다.
- SignUpViewModel의 주요 기능을 프로토콜로 추가했습니다.

song1900 committed:
Refactor: 주석 작성법을 수정했습니다.

주석 작성법을 통일시키고 퀵헬프에 제대로 표시 되도록 주석을 수정했습니다.

f-lab-jercy commented on GIL/GIL/Screen/SignUp/SignUpViewController.swift:
```diff
@@ -38,82 +38,203 @@ extension SignUpViewController {
+                    
+                case .failure(let error):
+                    // TODO: AlertService 적용 필요
+                    if let signUpError = error as? SignUpViewModel.SignUpError,
```
조금 더 명확하게 하기 위해서 여기서도 switch로 분기를 쳐보는건 어떨까요?

song1900 replied:
https://github.com/f-lab-edu/gil-navi-ios/pull/16/commits/eff258ce13d5a1375b8a085f418ba7d195242c96
해당 커밋으로 수정했습니다!

song1900 committed:
Merge branch 'main' of https://github.com/f-lab-edu/gil-navi-ios into feature/signup-registration

# Conflicts:
#	GIL/GIL.xcodeproj/project.pbxproj

song1900 committed:
Refactor: SignUp ViewController 코드 리팩토링

- createUserPublisher 에러 발생시 Switch문을 이용해 명확하게 분기처리
- AlertService 적용
- Style Guide 통일
